records/2021-09-27.batch-celery-save.md

```diff
@@ -0,0 +1,116 @@
+2- The integration is not that easy with Flask (although probably not impossible) as described by Flask integration recipe [in their documentation](https://flask.palletsprojects.com/en/2.0.x/appcontext/). [An issue was opened](https://github.com/clokep/celery-batches/issues/35) for curiosity sake to see if that could be done but even then, the issue described at #1 is a blocker.
+
+### Send notifications in queue and sweep process from a heartbeat process
+
+> **INFO**: There is a [Hearbeat](https://docs.celeryproject.org/en/stable/internals/reference/celery.worker.heartbeat.html) service within Celery that the latter manages. This is a one and only execution thread which allows to orchestrate certain tasks based on a schedule. This is useful as we can leverage a memory model that we will not share. Hence we avoid the pitfalls of concurrent execution thread trying to access shared memory all at the same time and which requires complex locking mechanism. This is similar to the UI model with an unique UI thread and its exclusive access to UI data, or the actor model which has one actor keeping its one exclusive state kept away from other actors.
+
+In order to achieve batch save, we need to buffer and retrieve the items from the
+Celery task. There is unfortunately no known mechanism at this time of writing of
+Celery that allows the retrieval of all items for a queue and process it as a batch
+(other than the previously mentioned and defunct celery-batch experimental library).
+
+There are [Celery primitives that allows to chain, iterate concurrently and group](https://docs.celeryproject.org/en/stable/userguide/canvas.html#the-primitives)
+which will be necessary to achieve a custom solution but no aggregate operation is
+available.
+
+Hence, we will need to buffer our notifications into a queue or cache system when
+we receive the notifications, prior to save these.
+
+![Buffer workflow of a notification save](https://raw.githubusercontent.com/cds-snc/notification-adr/main/records/diagrams/2021-09-27.batch-celery-save/buffer-workflow.jpg?raw=true "Buffer workflow of a notification save")
+
+We can do so by either storing the notifications in a queue (such as AWS SQS) or a
+distributed cache (such as AWS Elasticache Redis). Once these are stored, the
+heartbeat celery execution thread could check the stored notifications every
+second with a maximal number of notifications with an arbitrary number at first
+(i.e. 500 notifications) and properly weighted after performance tests runs.
+
+The `bulk_save_email` Celery task is new as this would bulk insert the notifications
+into the database but it would leverage the existing code and tasks. It would chain
+the newly created bulk notifications into the Celery tasks like the current workflow
+does; nothing new on that side.
+
+We should then consider which data store to use as a buffer for the Celery beat
```

Comment:
nit: s/beat/heartbeat

Reply:
Hmm that made me realize that Celery beat is actually the right nomenclature. I'll switch the other references to this one.

Reply:
Ah no wait, Celery actually uses the 2 in their documentation. I'll use heartbeat everywhere then to stick to one.